src/NuGet.Core/NuGet.Protocol/Plugins/Connection.cs

```diff
@@ -0,0 +1,263 @@
+        public ConnectionState State => (ConnectionState)_state;
+
+        /// <summary>
+        /// Occurs when an unrecoverable fault has been caught.
+        /// </summary>
+        public event EventHandler<ProtocolErrorEventArgs> Faulted;
+
+        /// <summary>
+        /// Occurs when a message has been received.
+        /// </summary>
+        public event EventHandler<MessageEventArgs> MessageReceived;
+
+        /// <summary>
+        /// Gets the message dispatcher.
+        /// </summary>
+        public IMessageDispatcher MessageDispatcher { get; }
+
+        /// <summary>
+        /// Gets the connection options
+        /// </summary>
+        public ConnectionOptions Options { get; }
+
+        /// <summary>
+        /// Gets the negotiated protocol version, or <c>null</c> if not yet connected.
+        /// </summary>
+        public SemanticVersion ProtocolVersion { get; private set; }
+
+        /// <summary>
+        /// Instantiates a new instance of the <see cref="Connection" /> class.
+        /// </summary>
+        /// <param name="dispatcher">A message dispatcher.</param>
+        /// <param name="sender">A sender.</param>
+        /// <param name="receiver">A receiver.</param>
+        /// <param name="options">Connection options.</param>
+        /// <exception cref="ArgumentNullException">Thrown if <paramref name="dispatcher" /> is <c>null</c>.</exception>
+        /// <exception cref="ArgumentNullException">Thrown if <paramref name="sender" /> is <c>null</c>.</exception>
+        /// <exception cref="ArgumentNullException">Thrown if <paramref name="receiver" /> is <c>null</c>.</exception>
+        /// <exception cref="ArgumentNullException">Thrown if <paramref name="options" /> is <c>null</c>.</exception>
+        public Connection(IMessageDispatcher dispatcher, ISender sender, IReceiver receiver, ConnectionOptions options)
+        {
+            if (dispatcher == null)
+            {
+                throw new ArgumentNullException(nameof(dispatcher));
+            }
+
+            if (sender == null)
+            {
+                throw new ArgumentNullException(nameof(sender));
+            }
+
+            if (receiver == null)
+            {
+                throw new ArgumentNullException(nameof(receiver));
+            }
+
+            if (options == null)
+            {
+                throw new ArgumentNullException(nameof(options));
+            }
+
+            MessageDispatcher = dispatcher;
+            _sender = sender;
+            _receiver = receiver;
+            Options = options;
+            _closeEvent = new TaskCompletionSource<object>();
+
+            MessageDispatcher.SetConnection(this);
+        }
+
+        /// <summary>
+        /// Disposes of this instance.
+        /// </summary>
+        public void Dispose()
+        {
+            if (_isDisposed)
+            {
+                return;
+            }
+
+            CloseAsync().GetAwaiter().GetResult();
```

Comment:
This is highly unrecommended practice as it would lead to a deadlock potentially.